File: app/web/src/locales/supportedLocales.ts
export interface LocaleInfo {
  bcp47: string;
  displayName: string;
  direction: "ltr" | "rtl";
}

export const supportedLocales: readonly LocaleInfo[] = [
  { bcp47: "en", displayName: "English (US)", direction: "ltr" },
  { bcp47: "en-GB", displayName: "English (UK)", direction: "ltr" },
  { bcp47: "es-419", displayName: "Español (Latinoamérica)", direction: "ltr" },
  { bcp47: "ca", displayName: "Català", direction: "ltr" },
  { bcp47: "zh-Hans", displayName: "简体中文", direction: "ltr" },
  { bcp47: "ar", displayName: "العربية", direction: "rtl" },
  { bcp47: "fr", displayName: "Français", direction: "ltr" },
  { bcp47: "hi", displayName: "हिन्दी", direction: "ltr" },
  { bcp47: "bn", displayName: "বাংলা", direction: "ltr" },
  { bcp47: "id", displayName: "Bahasa Indonesia", direction: "ltr" },
  { bcp47: "ru", displayName: "Русский", direction: "ltr" },
] as const;

export function findLocale(languageCode: string): LocaleInfo {
  const exact = supportedLocales.find((l) => l.bcp47 === languageCode);
  if (exact) return exact;

  const prefix = languageCode.split("-")[0];
  const prefixMatch = supportedLocales.find((l) => l.bcp47.startsWith(prefix));
  return prefixMatch ?? supportedLocales[0];
}

File: app/NeurOneShared/Sources/NeurOneShared/SupportedLocales.swift
import Foundation

public enum SupportedLocale: String, CaseIterable, Sendable {
    case en = "en"
    case enGB = "en-GB"
    case es419 = "es-419"
    case ca = "ca"
    case zhHans = "zh-Hans"
    case ar = "ar"
    case fr = "fr"
    case hi = "hi"
    case bn = "bn"
    case id = "id"
    case ru = "ru"

    public var bcp47Tag: String { rawValue }

    public var displayName: String {
        switch self {
        case .en: return "English (US)"
        case .enGB: return "English (UK)"
        case .es419: return "Español (Latinoamérica)"
        case .ca: return "Català"
        case .zhHans: return "简体中文"
        case .ar: return "العربية"
        case .fr: return "Français"
        case .hi: return "हिन्दी"
        case .bn: return "বাংলা"
        case .id: return "Bahasa Indonesia"
        case .ru: return "Русский"
        }
    }

    public var isRTL: Bool {
        self == .ar
    }

    public static func from(languageCode: String) -> SupportedLocale {
        if let exact = SupportedLocale(rawValue: languageCode) {
            return exact
        }
        let prefix = languageCode.components(separatedBy: "-").first ?? languageCode
        switch prefix {
        case "es": return .es419
        case "zh": return .zhHans
        default:
            return SupportedLocale.allCases.first { $0.rawValue.hasPrefix(prefix) } ?? .en
        }
    }
}
